Sources/PluginLibrary/NamingUtils.swift

```diff
@@ -198,90 +198,74 @@ fileprivate func sanitizeTypeName(_ s: String, disambiguator: String) -> String
 }
 
 fileprivate func isCharacterUppercase(_ s: String, index: Int) -> Bool {
-  let start = s.index(s.startIndex, offsetBy: index)
-  if start == s.endIndex {
+  let scalars = s.unicodeScalars
+  let start = scalars.index(scalars.startIndex, offsetBy: index)
+  if start == scalars.endIndex {
     // it ended, so just say the next character wasn't uppercase.
     return false
   }
-  let end = s.index(after: start)
-  let sub = s[start..<end]
-  return sub != sub.lowercased()
+  return scalars[start].isUppercase
+}
+
+fileprivate func makeUnicodeScalarView(
+  from unicodeScalar: UnicodeScalar
+) -> String.UnicodeScalarView {
+  var view = String.UnicodeScalarView()
+  view.append(unicodeScalar)
+  return view
 }
 
-fileprivate let digits: Set<String> = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
 
 fileprivate func splitIdentifier(_ s: String) -> [String] {
-  var out = [String]()
-  var current = ""
-  var last = ""
+  var out: [String.UnicodeScalarView] = []
+  var current = String.UnicodeScalarView()
+  var last: UnicodeScalar = "\0"
   var lastIsUpper = false
   var lastIsLower = false
 
-  for _c in s.characters {
-    let c = String(_c)
-    let cIsUpper = (c != c.lowercased())
-    let cIsLower = (c != c.uppercased())
-    if digits.contains(c) {
-      if digits.contains(last) {
-        current += c
+  for scalar in s.unicodeScalars {
+    let isUpper = scalar.isUppercase
+    let isLower = scalar.isLowercase
+
+    if scalar.isDigit {
+      if last.isDigit {
+        current.append(scalar)
       } else {
         out.append(current)
-        current = c
+        current = makeUnicodeScalarView(from: scalar)
       }
-    } else if cIsUpper {
+    } else if isUpper {
       if lastIsUpper {
-        current += c.lowercased()
+        current.append(scalar.lowercased())
       } else {
         out.append(current)
-        current = c.lowercased()
+        current = makeUnicodeScalarView(from: scalar.lowercased())
       }
-    } else if cIsLower {
+    } else if isLower {
       if lastIsLower || lastIsUpper {
-        current += c
+        current.append(scalar)
       } else {
         out.append(current)
-        current = c
-      }
-    } else {
-      if last == "_" {
-        out.append(current)
-        current = last
-      }
-      if c != "_" {
-        out.append(current)
-        current = c
+        current = makeUnicodeScalarView(from: scalar)
       }
+    } else if last == "_" {
+      out.append(current)
+      current = makeUnicodeScalarView(from: last)
```

Comment:
What happens on an invalid character, it gets dropped, but does it get the states right?  i.e. before what happened to "abc$xyz" and what happens now?

As long as we only ever send field names, enums, etc. here it shouldn't happen.  Just not sure if we want to trap it in some way as an added safety net.

Reply:
I just tested it and characters outside a-z, A-Z, 0-9, and _ still cause it to split correctly around them, because `lastIsUpper` and `lastIsLower` still get updated for every character, so we still split around them when we encounter the next character.